safedelete/managers.py

```diff
@@ -100,6 +101,32 @@ def all(self, **kwargs):
             qs._safedelete_force_visibility = force_visibility
         return qs
 
+    def update_or_create(self, defaults=None, **kwargs):
+        """Regular update_or_create() fails on soft-deleted, existing record with unique constraint on non-id field
+
+        Args:
+            defaults: Dict with defaults to update/create model instance with
+            kwargs: Attributes to lookup model instance with
+        """
+
+        # Check if one of the model fields contains a unique constraint
+        if self.model.has_unique_fields(self.model):
+            # Check if object is already soft-deleted
+            deleted_object = self.all_with_deleted().filter(**kwargs).exclude(deleted=None).first()
+
+            # If object is soft-deleted, reset delete-state...
+            if deleted_object and deleted_object.get_delete_policy() in self.get_soft_delete_policies():
+                deleted_object.deleted = None
+                deleted_object.save()
+
+        # Do the standard logic
+        return super(SafeDeleteManager, self).update_or_create(defaults, **kwargs)
+
+    @staticmethod
+    def get_soft_delete_policies():
```

Comment:
Unneeded.

Reply:
I need it to see if I'm in the soft-delete case. I think it's DRYer than putting the two constants in the if. I'd like to leave it the way it is.

Reply:
It's not pythonic to use getters, and using the method once doesn't warrant a property imo. I'd rather see it as a constant as well and not a method/static method.

Reply:
So how would you implement that I can query if a policy is soft?

Reply:
I think

```python
if deleted_object and deleted_object._safedelete_policy in [SOFT_DELETE, SOFT_DELETE_CASCADE]:
```

is fine here. It's not used elsewhere and it immediately makes clear what these soft delete policies are. @Gagaro got the final say here though.

Reply:
I agree with @AndreasBackx . If it is only used once, we should prefer readability.

Reply:
@Gagaro My idea is that you can use this function as well in your own project code. So if I want to know if my model has a soft delete policy, it's a useful helper function. I like to encapulate constants like this to avoid having any hazzle later on if at some point the constants themself change or one is added or removed. Won't you agree?